### PR TITLE
Option to change number of required artifacts

### PR DESCRIPTION
**Pull request by gollop**

### CODE REVIEW

File: src/patches.rs [hunk 1]
```diff
@@ -13702,6 +13706,42 @@ fn build_and_run_patches(gc_disc: &mut structs::GcDisc, config: &PatchConfig, ve
...
+    if artifact_count > 12 {
+        panic!("Must specify between 0 and 12 required artifacts");
+    }
+
+    patcher.add_scly_patch(
+        resource_info!("07_stonehenge.MREA").into(),
+        move |_patcher, area| {
+            let scly = area.mrea().scly_section_mut();
+
+            let layer = &mut scly.layers.as_mut_vec()[1]; // Monoliths and Ridley
```
Comment (toasterparty) [inline src/patches.rs]: Ulti has a function for getting layers by name now, we should probably use it

**Review by toasterparty — APPROVED** [review]

File: src/patches.rs (hunk 1 above)
Comment (gollop) [reply src/patches.rs]: Done

**Comment by toasterparty** [comment]
The implications of this are a little confusing. Consider the following case:
- 6 Artifacts are placed in the world
- You use this new config option to set 3 required artifacts
- randomprime then (automatically) fills in the 9 unplaced artifacts with already activated relays

Would the artifact temple then immediately open upon first visit? Or do you actually still need to find 3 artifacts?

**Comment by toasterparty** [comment]
Also, how does this interact with the `artifact_temple_layer_overrides` option used by multiworld?

**Comment by gollop** [comment]
> Also, how does this interact with the `artifact_temple_layer_overrides` option used by multiworld?

No idea. Untested and not sure how that works

**Comment by gollop** [comment]
> The implications of this are a little confusing. Consider the following case:
> 
>     * 6 Artifacts are placed in the world
> 
>     * You use this new config option to set 3 required artifacts
> 
>     * randomprime then (automatically) fills in the 9 unplaced artifacts with already activated relays
> 
> 
> Would the artifact temple then immediately open upon first visit? Or do you actually still need to find 3 artifacts?

This PR doesn't work like that. It changes a single counter, no changes are made to memory relays.

**Comment by toasterparty** [comment]
> randomprime then (automatically) fills in the 9 unplaced artifacts with already activated relays

This bit is describing existing functionality. I would have assumed that the counter starts at 9 the first time you enter artifact temple in these cases.

**Comment by Antidote** [comment]
If those layers are set to active by default it would trigger the ridley fight right away if the counter value is changed.

**Comment by gollop** [comment]
> > randomprime then (automatically) fills in the 9 unplaced artifacts with already activated relays
> 
> This bit is describing existing functionality. I would have assumed that the counter starts at 9 the first time you enter artifact temple in these cases.

It would be 6 unplaced artifacts, wouldn't it?

**Comment by toasterparty** [comment]
If what Antidote says is true, then this PR should set the counter to something like:
```
counter = 12 - (desired_artifact_count - placed_artifact_count)
```

**Comment by Antidote** [comment]
> If what Antidote says is true, then this PR should set the counter to something like:
> ```
> counter = 12 - (desired_artifact_count - placed_artifact_count)
> ```

This is correct, since each layer has a relay that increments the counter when it's activated.

**Comment by gollop** [comment]
> > If what Antidote says is true, then this PR should set the counter to something like:
> > ```
> > counter = 12 - (desired_artifact_count - placed_artifact_count)
> > ```
> 
> This is correct, since each layer has a relay that increments the counter when it's activated.

The counter is decremented, not incremented

**Comment by toasterparty** [comment]
The math is the same for increment/decrement

**Review by toasterparty — CHANGES_REQUESTED** [review]
Please show that 6 placed artifacts + 3 required artifacts works as one would expect

**Comment by gollop** [comment]
Minimal example seed

> {
>   "forceVanillaLayout": false,
>   "externAssetsDir": null,
>   "seed": 1,
>   "preferences": {
>     "qolCutscenes": "Major",
>     "qolGameBreaking": true,
>     "qolCosmetic": true,
>     "qolPickupScans": true
>   },
>   "gameConfig": {
>     "removeVanillaBlastShields": true,
>     "warpToStart": true,
>     "requiredArtifactCount": 11,
>     "startingItems": {
>       "combatVisor": true,
>       "powerBeam": true,
>       "scanVisor": true,
>       "missiles": 5,
>       "energyTanks": 0,
>       "powerBombs": 0,
>       "wave": false,
>       "ice": false,
>       "plasma": false,
>       "charge": true,
>       "morphBall": true,
>       "bombs": false,
>       "spiderBall": false,
>       "boostBall": false,
>       "variaSuit": false,
>       "gravitySuit": false,
>       "phazonSuit": false,
>       "thermalVisor": false,
>       "xray": false,
>       "spaceJump": true,
>       "grapple": false,
>       "superMissile": false,
>       "wavebuster": false,
>       "iceSpreader": false,
>       "flamethrower": false
>     },
>     "startingRoom": "tallon:Landing Site"
>   },
>   "tweaks": {},
>   "levelData": {
>     "Tallon Overworld": {
>       "transports": {
>         "Tallon Overworld East\u0000(Frigate Crash Site)": "chozo:Transport to Tallon Overworld East",
>         "Tallon Overworld South\u0000(Great Tree Hall, Upper)": "chozo:Transport to Tallon Overworld South",
>         "Tallon Overworld North\u0000(Tallon Canyon)": "chozo:Transport to Tallon Overworld North",
>         "Artifact Temple": "crater:Crater Entry Point",
>         "Tallon Overworld West\u0000(Root Cave)": "magmoor:Transport to Tallon Overworld West",
>         "Tallon Overworld South\u0000(Great Tree Hall, Lower)": "mines:Transport to Tallon Overworld South"
>       },
>       "rooms": {
>         "Frigate Crash Site": {
>           "pickups": [
>             {
>               "pickupType": "Artifact of Nature",
>               "currIncrease": 5,
>               "maxIncrease": 5,
>               "model": "Artifact of Nature",
>               "scanText": "Artifact of Nature",
>               "hudmemoText": "Artifact of Nature acquired!",
>               "destination": null
>             }
>           ],
>           "doors": {}
>         },
>         "Alcove": {
>           "pickups": [
>             {
>               "pickupType": "Artifact of Spirit",
>               "currIncrease": 5,
>               "maxIncrease": 5,
>               "model": "Artifact of Spirit",
>               "scanText": "Artifact of Spirit",
>               "hudmemoText": "Artifact of Spirit acquired!",
>               "destination": null
>             }
>           ],
>           "doors": {}
>         }
>       }
>     }
>   }
> }

10 artifacts not in world, 11 required artifacts. Collect no artifacts, no Ridley fight. Collect either of the two artifacts in the world, Ridley fight occurs.

**Pull request merged**